### PR TITLE
docs: frontend decisions from meeting 2019-04-11

### LINKED ISSUE #15: One file for one function
I know that this is a topic where MOST devs will not agree with my opinion to put every function into a separate file. I don't want to convince anyone, I just want to share my experience with both ways and want to hear your opinion about it.

# Benefits of using one file per function

### Testing functions
This is the by far more important part of separating functions into separate files.
Let's consider the following example:

```js
// fooBarBaz.js
export const foo = () => setTimeout(() => console.log('foo'), 100);
export const bar = () => setTimeout(() => console.log('bar'), 200);
export const baz = () => setTimeout(() => console.log('baz'), 300);

export const fooBarBaz = toggle => {
    foo()
    if (toggle) {
        bar();
    } else {
        baz();
    }
}
```

Testing the `fooBarBaz` function will result in checking whether the right things have been logged.
I can't stub the three functions that are being called inside `fooBarBaz` because they're in the same module. I know that I can disable timers to make this task easier, but that's not my point, just imagine the three functions do something  of async nature.

**What's the problem with testing the `console.log`s here?**
The problem is that I test the three functions twice. If there are tests in place for the three functions I don't care about what they do in the test for `fooBarBaz`, I only want to test whether the right functions are called for specific scenarios. Imagine I have multiple functins inside this file which all rely on the three functions, for every function that I test, I have to test the output of `foo`, `bar` and `buz` again!

Ideally my test would look like this:
```js
// __test__/fooBarBaz.test.js
import { fooBarBaz } from '../fooBarBaz'
import { foo } from '../fooBarBaz/foo'
import { bar } from '../fooBarBaz/bar'
import { baz } from '../fooBarBaz/baz'

jest.mock('../fooBarBaz/foo', () => ({ foo: jest.fn() }))
jest.mock('../fooBarBaz/bar', () => ({ bar: jest.fn() }))
jest.mock('../fooBarBaz/baz', () => ({ baz: jest.fn() }))

describe('fooBarBaz', () => {
    it('should call foo and bar when called with true', () => {
        fooBarBaz(true);
        expect(foo).toHaveBeenCalledTimes(1);
        expect(bar).toHaveBeenCalledTimes(1);
        expect(baz).not.toHaveBeenCalled();
    });

    it('should call foo and baz when called with false', () => {
        fooBarBaz(false);
        expect(foo).toHaveBeenCalledTimes(1);
        expect(baz).toHaveBeenCalledTimes(1);
        expect(bar).not.toHaveBeenCalled();
    });1
})
```

This way I really don't care whether the three functions return the right stuff in this test. Which is awesome, because this test isn't about the three functions anyways! **Imagine testing thunks like this!**

### Fuzzy search
When looking for a function, it's way easier to find the right file without looking through the project's file tree by just entering the file name and getting a list of files with exactly that name.

**Why is this important to me?**
When programming I don't want to spend time with the tooling. Ideally I wouldn't even spend time programming at all if I could create the solution in a faster way. So my final goal is to reduce the time I spend working on something to get it done asap.

Let's say looking for a file containing a function I'm looking for takes 3 seconds longer than just typing the functions name and hitting enter. The more complex (or complicated; look at the maintenance app) a project is, the more I will be looking for files, so if I end up looking for a function 100 times (which is not an unrealistic figure given that we're working 8 hours per day!), it's 5 minutes per day or 25 Minutes per week I spend extra which I wouldn't need to spend if looking for functions was fuzzy search friendly.

It doesn't sound like a lot and what I want to say is that there are MANY things that can be improved and this is one of those. Anyways it might take A LOT LONGER to find a function than 3 seconds if they're hidden behind an abstract file name.

### Meaningful names

If you separate functions into separate files, you'll have to think a bit more about the function's names. This will make the code more declarative, which is always good! It's a bit like TDD, it forces you to think about stuff.

That might slow you down, but in the long run, having proper function names can save a lot of time. This is basically the same reason why spending extra time for unit tests is totally ok.

## My approach to not get a messy codebase

I think this is what most people hate about separating functions into many files and I have to say it takes a bit of discipline here.

Functions that are used only inside a specific function can be outsourced to make code declarative (which is something we really want in functional programming), so reading through the code looks more like reading through a set of instructions until we read the code of the low level units. When done right, you never touch these low level units until you either need to know the implementation or you're just curious. And you only need to touch the implementation if they are the cause for bugs, otherwise you really won't have to deal with them.

So what I do is:
I extract logic from a function, create a new function with a meaningful name, describing WHAT is happening and now HOW something is done (because that's the imperative part of the code that we want to hide as much as possible), then I create a folder with the name of the function I extracted code from. In the example above, I'd create a `fooBarBaz` folder and put the three functions inside there.
I'd never have to look at the three functions until I need to know the implementation details AND they don't clutter the folder where `fooBarBaz` is inside.

Functions that are meant to be reused I always put inside a proper namespace that's inside `src/utils` or `src/lib` or `src/helpers` (personal preference, I prefer `lib` as it's faster to write).

## Conclusion
We want our code to be tested and that's good. Ideally we don't want to rewrite our code to be testable, so it makes sense to write code in a way that it feels like a "natural" code structure while still being testable. 
I never had problems with finding files with fuzzy search and I was able to test functions without having to test other functions functionality again. Which can be a tedious process when refactoring!

So what are your thoughts about this? Have you ever followed the same approach but found yourself lost in the amount of files?

## Update
This doesn't mean that I put ALL files into a separate file. Especially view files like react components often have multiple stuff inside one file in my code. Also after we've reached the conclusion that you shouldn't test EVERYTHING just for the sake of code coverage, it makes sense to group code that doesn't really need only needs test for like one or two exported functions of the whole module.

### LINKED ISSUE #23: Which future syntax do we want to use in our js?
I was going to open a jira issue for this, but it seemed slightly too philosophical for that. I'm wondering what level of the ecmascript proposals we officially want to use in our apps. As in, do we allow any and all future syntax? Do we only use syntax that's highly likely to not have to be changed?

I'm asking because we're using some future syntax in our apps, but I'm not aware of any dhis2 stance on what we do and don't use. So it seems to be up to the app devs. I think it would be good to have some clarity on that by sharing/discussing our stance.

For reference, the stages:

> Stage 0 “strawman” — The starting point for all proposals. These can change significantly before advancing to the next stage. There is no acceptance criteria and anyone can make a new proposal for this stage. There doesn’t need to be any implementation and the spec isn’t held to any standard. This stage is intended to start a discussion about the feature. There are currently over twenty stage 0 proposals.

> Stage 1 “proposal” — An actual formal proposal. These require a “champion”(i.e. a member of TC39 committee). At this stage the API should be well thought out and any potential implementation challenges should be outlined. At this stage, a polyfill is developed and demos produced. Major changes might happen after this stage, so use with caution. Proposals at this stage include the long-awaited Observables type and the Promise.try function.

> Stage 2 “draft” — At this stage the syntax is precisely described using the formal TC39 spec language. Minor editorial changes might still happen after this stage, but the specification should be complete enough not to need major revisions. If a proposal makes it this far, it’s a good bet that the committee expects the feature to be included eventually.

> Stage 3 “candidate” — The proposal has approved and further changes will only occur at the request of implementation authors. Here is where you can expect implementation to begin in JavaScript engines. Polyfills for proposals at this stage are safe to use without worry.

> Stage 4 “finished” — Indicates that the proposal has been accepted and the specification has been merged with the main JavaScript spec. No further changes are expected. JavaScript engines are expected to ship their implementations. As of October 2017 there are nine finished proposals, most notably async functions.

> From https://blog.logrocket.com/interesting-ecmascript-2017-proposals-163b787cf27c

Related to this we might also have to discuss APIs actually.

### LINKED ISSUE #31: Validation of config objects
So, discussing a PR with @Mohammer5 about a function with many parameters we decided to go with a config object instead.

E.g.
```
/**
* @param {number} a - the first number
* @param {number} b - etc....
*/
const fn = (a, b, c, d, e, f, g) => {
}
```
goes into
```
/**
* @param {object} config I have all the data!
* @param {number} config.a The first number....
* @param {number} config.b etc....
*/
const fn = ( config ) => { 
    const { a, b, c, d, e, f, g } = config
}
```

Then the discussion took fire and our thoughts went apart so we decided that a notes would be good to ventilate it with the team.

The 2 different takes at this is the following:

1. Strict validation of config objects either manually or with a Joi schema - **not allowing unknowns**
```
const def = joi.object().keys({
    a: joi.number().required(),
    b: joi.number().required()
})
```
This would make it **fail** if trying to pass `{ a: 1, b: 2, c: 3 }` because **c is not allowed**.

Instead if we'd like to allow c but have it optional we have to specify it as such.
```
const def = joi.object().keys({
    a: joi.number().required(),
    b: joi.number().required(),
    c: joi.number().optional
})
```

2. Loose validation, validating the specified schema, but **allowing unknown properties** 

This is easy to implement with Joi by just adding `stripUnknown: true` to the schema. The question is rather which of them we want to use. The 2'nd option would make the example pass for property `c` as unknown even if not specified as optional.

Personally I'd like to know if I pass an object with incorrect parameters even if they are unknown. For me, passing data to a function actually means that I expect them to be used (except optionals), not that it should be up to the function to swallow huge objects and use whatever they want to. So I'm for number 1

@Mohammer5 is on the other side and would like the second one.

Maybe we can set a standard for this in terms of both if we should use config objects where the amount of parameters exceed a specific count, and how they should be validated. Should we even use Joi (except in the app store backend where its a part for testing hapijs)? - or something else.

### PR DESCRIPTION
**Pull request by varl**
Thanks everyone for a constructive meeting. I hope the format makes sense and isn't _too_ arbitrary. Suggestions on how to improve the process is always welcome.

Please review the meeting notes and hit the **Approve** button if it looks alright.

If you feel that something is incorrect or that something is misrepresented in the notes from what was discussed, hit **Request changes**.

If you have any suggestions on how to make the meeting minute better and more clear, please add a review with comments!

Closes #23
Closes #15
Closes #31

### CODE REVIEW

**Review by ghost — APPROVED** [review]

**Comment by ismay** [comment]
Not sure how the 'on behalf of' got there, but I assume that's ok:

Screenshot 2019-04-11 at 17 51 44

**Comment by erikarenhill** [comment]
I'd like to allow a maximum of 3 positional parameters instead of less than 3, otherwise LGTM!

**Review by Mohammer5 — APPROVED** [review]
:+1: Very good!
@ismay It's the same for me.. I guess it'll be the same for everyone?

**Review by JoakimSM — APPROVED** [review]
👍

**Review by Birkbjo — APPROVED** [review]
I'm with @erikarenhill, three and below may use positional arguments. Very minor and just preference.
Other than that :+1:

File: decisions/2019/04/11.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+## #31 Validation of config objects
+
+**Conclusion**: This is inherently a problem that depends on the
+context.
+
+The general rule for parameters is the relative standard:
+
+- Less than 3 params can be positional
+- 3 and above warrants a configuation object.
```
Comment (HendrikThePendric) [inline decisions/2019/04/11.md]: Less than 4 params can be positional
4 and above warrants a configuration object.

**Review by erikarenhill — APPROVED** [review]

File: decisions/2019/04/11.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+developers and reviewers. It can have a big impact on understanding a
+code base, so do keep it in mind to err on the side of readability.
+Files are cheap as Git only tracks content.
+
+Generally smaller files are better. Use common sense as to what makes
+the code easier to read and understand.
+
+Utility files is one example where if every util function is kept in the
+same file it becomes a mess and hard to read and keep track of.
+Divide and conquer.
```
Comment (HendrikThePendric) [inline decisions/2019/04/11.md]: Optionally add a sentence like this:
> Never export more than one independent component from a single file

(I mean components that are actually meant be used independently, it's OK if some internal/helper component is exported from a file for unit-testing)

**Review by HendrikThePendric — COMMENTED** [review]
Like others have already stated, please param++

File: decisions/2019/04/11.md (hunk 1 above)
Comment (varl) [reply decisions/2019/04/11.md]: I've updated to **max 3 positional arguments**.

**Review by HendrikThePendric — APPROVED** [review]

**Comment by varl** [comment]
>I'd like to allow a maximum of 3 positional parameters instead of less than 3, otherwise LGTM!

>I'm with @erikarenhill, three and below may use positional arguments.

>Like others have already stated, please param++

I did some quick scrapes of our code base and **max 3** seems to make sense. Updated.

**Pull request merged**